agent/xds/listeners.go

```diff
@@ -200,31 +208,58 @@ func (s *Server) listenersFromSnapshotMeshGateway(cfgSnap *proxycfg.ConfigSnapsh
 			addr = "0.0.0.0"
 		}
 
-		l, err := s.makeGatewayListener("default", addr, cfgSnap.Port, cfgSnap)
-		if err != nil {
-			return nil, err
+		a := structs.ServiceAddress{
+			Address: addr,
+			Port:    cfgSnap.Port,
+		}
+		if !seen[a] {
+			addrs = append(addrs, namedAddress{name: "default", ServiceAddress: a})
+			seen[a] = true
 		}
-		resources = append(resources, l)
 	}
 
 	if cfg.BindTaggedAddresses {
 		for name, addrCfg := range cfgSnap.TaggedAddresses {
-			l, err := s.makeGatewayListener(name, addrCfg.Address, addrCfg.Port, cfgSnap)
-			if err != nil {
-				return nil, err
+			a := structs.ServiceAddress{
+				Address: addrCfg.Address,
+				Port:    addrCfg.Port,
+			}
+			if !seen[a] {
+				addrs = append(addrs, namedAddress{name: name, ServiceAddress: a})
+				seen[a] = true
 			}
-			resources = append(resources, l)
 		}
 	}
 
 	for name, addrCfg := range cfg.BindAddresses {
-		l, err := s.makeGatewayListener(name, addrCfg.Address, addrCfg.Port, cfgSnap)
-		if err != nil {
-			return nil, err
+		a := structs.ServiceAddress{
+			Address: addrCfg.Address,
+			Port:    addrCfg.Port,
+		}
+		if !seen[a] {
+			addrs = append(addrs, namedAddress{name: name, ServiceAddress: a})
+			seen[a] = true
 		}
-		resources = append(resources, l)
 	}
 
+	// Make listeners once deduplicated
+	for _, a := range addrs {
+		var l *envoy.Listener
+
+		switch cfgSnap.Kind {
+		case structs.ServiceKindTerminatingGateway:
+			l, err = s.makeTerminatingGatewayListener(a.name, a.Address, a.Port, cfgSnap, token)
+			if err != nil {
+				return nil, err
+			}
+		case structs.ServiceKindMeshGateway:
+			l, err = s.makeMeshGatewayListener(a.name, a.Address, a.Port, cfgSnap)
+			if err != nil {
+				return nil, err
+			}
+		}
+		resources = append(resources, l)
```

Comment:
Do you want to guard this append in the event a bad kind comes through and `l` ends up being `nil`?

Reply:
Yep, fixed!